src/julia-syntax.scm

```diff
@@ -1646,8 +1647,9 @@
                  ,@(if (and (not outer?) (or *depwarn* *deperror*))
                        (map (lambda (v) `(warn-if-existing ,v)) (lhs-vars lhs))
                        '())
-                 ,(lower-tuple-assignment (list lhs state)
-                                          `(call (top next) ,coll ,state))
+                 (= ,nxt (call (top next) ,coll ,state))
+                 (if (call (core ===) ,nxt (core nothing)) (break))
```

Comment:
I assume this check gets optimized out in the common case of a `next` function that always returns a tuple?

Hopefully it gets optimized out even for iteration over a type-unstable container, e.g. where `next` returns `Tuple{ ...inference fails... , MyState}`?